### DISCUSSION TITLE
Combine redux and usePouchDb

### DISCUSSION THREAD

**Discussion by jnnwnk** [Terreii/use-pouchdb#38] [Q&A]
This is actually more of a question about architecture and it doesn't really feel like this right place for it.
Anyway, maybe it's still helpful for others or the documentation can be extended accordingly.

The documentation (https://christopher-astfalk.de/use-pouchdb/docs/api/provider#with-redux-and-react-redux) briefly points out that a combination of `Redux` and `usePouchDb` is possible.
I'm not quite sure how to split it up in a reasonable way. The hooks of `usePouchDb` make it easy to access the data in all components. The same would also work with `Redux`. So my feeling was that the use of `usePouchDb` and `Redux` is not really necessary and that the decision for one approach is sufficient. But this does not seem to be the case. 

Does anyone have an idea for a useful architecture? I would be very interested in ideas and opinions. :)

**Comment by Terreii** [comment]
Sorry that I didn't answer sooner! 🙇

This answer also applies to the use of [react-pouchdb](https://www.npmjs.com/package/react-pouchdb). React-pouchdb is like usePouchDB, but it uses [`<Suspense />`](https://reactjs.org/docs/react-api.html#suspense) and [Error Boundaries](https://reactjs.org/docs/error-boundaries.html#gatsby-focus-wrapper).

## Why you would require both

The example was more an reminder, that you can use both. I added it to show a possible solution. I personally have never used [redux-pouchdb](https://www.npmjs.com/package/redux-pouchdb) or any other Redux and PouchDB connection packages.
While I would recommend to start with hooks, there are some situation where you could require both (use-pouchdb and redux):
- While you are transition from one to the other.
- You have derived data ([reselect](https://www.npmjs.com/package/reselect)) or more complex relationships.
- And/or they are required in multiple components.
- You don't have a local DB and controlling the amount of network traffic is imported.

Redux is good for shared state. You are also more in control of it.

I have for example a project where a UDP connection (with Web-Sockets on the web) is established to another service. While most of the state will live in the Redux store, some of it will be stored in PouchDB. I then can use usePouchDB to display it.

## The architecture
There are two possible solution I can think of for now:
- One where Redux controls PouchDB DBs.
- And one where React controls PouchDB DBs (this also works with [react-pouchdb](https://www.npmjs.com/package/react-pouchdb))

With control I mean creation and destruction of DBs. This is also probably where you would check your session state (is the user logged in?).

Both examples are with [redux-thunk](https://www.npmjs.com/package/redux-thunk).

### Redux is in control
You have some actions for checking the session state, sign up, sign in and sign out. They would create a local DB and store it in the [extra argument of redux-thunk](https://www.npmjs.com/package/redux-thunk#injecting-a-custom-argument).

In your component that renders the [usePouchDB `<Provider />` component](https://christopher-astfalk.de/use-pouchdb/docs/api/provider) you would access the DB by a little hack:

```jsx
import React, { useState, useEffect } from 'react'
import { Provider } from 'use-pouchdb'
import { useDispatch, useSelector } from 'react-redux'

import { selectIsLoggedIn } from './selectors'

export function PouchDbContext ({ children }) {
  const dispatch = useDispatch()
  const [db, setDB] = useState(null)
  const isLoggedIn = useSelector(selectIsLoggedIn) // Or something else that changes with the DB.

  useEffect(() => {
    if (isLoggedIn) {
      // dispatch returns the result of the Thunk action.
      // The action only accesses the db from the extra argument and returns it.
      const nextDB = dispatch((dispatch, getState, { db }) => db)
      setDB(nextDB)
    } else {
      setDB(null) // or your new one, if you also have a local db if no user is logged in.
    }
  }, [dispatch, isLoggedIn])

  // Only render children and Provider if there is a db.
  if (db == null || !isLoggedIn) return null
  return <Provider pouchdb={db}>{children}</Provider>
}
```

It is imported, that all hooks have always access to a DB.

### React is in control

Here you have a component that creates and destroys the local db (like in the [Basic Tutorial](https://christopher-astfalk.de/use-pouchdb/docs/basics/provider)).
Then there is a hook, in that component or one of its children, that accesses the DB and dispatches an action if it changes.

```javascript
import { useEffect } from 'react'
import { usePouch } from 'use-pouchdb'
import { useDispatch } from 'react-redux'

export function useSetReduxDB () {
  const db = usePouch()
  const dispatch = useDispatch()

  useEffect(() => {
    // this thunk action accesses the extra argument and stores the db in it.
    // And then it notifies the redux-store of this change.
    dispatch((dispatch, getState, extraArg) => {
      extraArg.db = db
      dispatch({ type: 'DB_CHANGED' })
    })
  }, [db, dispatch])
}
```

It is also imported that some log running Redux-Thunk-Actions, that listen to DB-changes, unsubscribe. Dispatch an action before changing the DB on the extra argument, which does that. And/or let them listen to the [`destroyed`-event](https://pouchdb.com/api.html#events) of PouchDB.

P.S. I transformed this into a discussion. When the question was asked, discussions wasn't activated for this project.
